### PR TITLE
Work on #70 - Added configuration for defining measures and views

### PR DESCRIPTION
**Pull request by JonasKunz**

### CODE REVIEW

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManager.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        Measure measure;
+        switch (fullDefinition.getType()) {
+            case LONG:
+                measure = Measure.MeasureLong.create(fullDefinition.getName(), fullDefinition.getDescription(), fullDefinition.getUnit());
+                break;
+            case DOUBLE:
+                measure = Measure.MeasureDouble.create(fullDefinition.getName(), fullDefinition.getDescription(), fullDefinition.getUnit());
+                break;
+            default:
+                throw new RuntimeException("Unhandled case!");
```
Comment (mariusoe) [inline inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManager.java]: Let's use a more specific message "The type '{}' is not supported for measurements."

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManager.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                return instance instanceof Aggregation.Distribution;
+            case LAST_VALUE:
+                return instance instanceof Aggregation.LastValue;
+            default:
+                throw new RuntimeException("Unhandled case!");
+        }
+    }
+
+
+    private Aggregation getAggregationOfType(ViewDefinitionSettings.Aggregation aggregationType, List<Double> histogramBuckets) {
```
Comment (mariusoe) [inline inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManager.java]: Optional Comment:
how about changing the signatare to `getViewAggregation(ViewDefinitionSettings definitionSettings)`?

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManager.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            case COUNT:
+                return instance instanceof Aggregation.Count;
+            case SUM:
+                return instance instanceof Aggregation.Sum;
+            case HISTOGRAM:
+                return instance instanceof Aggregation.Distribution;
+            case LAST_VALUE:
+                return instance instanceof Aggregation.LastValue;
+            default:
+                throw new RuntimeException("Unhandled case!");
```
Comment (mariusoe) [inline inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManager.java]: `Aggregation type '{}' is not supported.`

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManager.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+            case COUNT:
+                return Aggregation.Count.create();
+            case SUM:
+                return Aggregation.Sum.create();
+            case HISTOGRAM:
+                return Aggregation.Distribution.create(BucketBoundaries.create(histogramBuckets));
+            case LAST_VALUE:
+                return Aggregation.LastValue.create();
+            default:
+                throw new RuntimeException("Unhandled case!");
```
Comment (mariusoe) [inline inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManager.java]: `Aggregation type '{}' is not supported.`

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/HookManager.java [hunk 5]
```diff
@@ -226,13 +227,6 @@ private void deactivateRemovedHooks(Class<?> clazz, Map<MethodDescription, Metho
...
     private void selfMonitorQueueSize() {
-        if (selfMonitoring.isSelfMonitoringEnabled()) {
-            val measure = selfMonitoring.getSelfMonitoringMeasureLong(
-                    "instrumentation-hooking-queue-size",
-                    "The number of pending classes inspectIT has to check if they require instrumentation updates",
-                    "classes",
-                    Aggregation.LastValue::create);
-            selfMonitoring.recordMeasurement(measure, pendingClasses.size());
-        }
+        selfMonitoring.recordMeasurement("instrumentation-hooking-queue-size", pendingClasses.size());
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/HookManager.java]: We will remove the name property and hardcode the inspectit/self/ prefix.

File: inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/system/ThreadMetricsRecorder.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    private static final String PEAK_METRIC_NAME = "peak";
+
+    private static final String DAEMON_METRIC_NAME = "daemon";
+
+    private static final String LIVE_METRIC_NAME = "live";
+
+    private static final String STATE_METRIC_NAME = "states";
+
+    private static final String STATE_TAG_NAME = "state";
+    private static final String METRIC_THREAD_UNIT = "threads";
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/metrics/system/ThreadMetricsRecorder.java]: UNIT is not used here anymore as the unit is defined in the default.yml now -> I removed the field

File: inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    @Mock
+    ViewManager viewManager;
+
+    @Mock
+    CommonTagsManager commonTagsManager;
+
+    @Mock
+    Aggregation aggregation;
+
+    private final Supplier<Aggregation> aggregationSupplier = () -> aggregation;
```
Comment (mariusoe) [inline inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManagerTest.java]: Not used

File: inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+public class MeasuresAndViewsManagerTest {
+
+    @Mock
+    ViewManager viewManager;
+
+    @Mock
+    CommonTagsManager commonTagsManager;
+
+    @Mock
+    Aggregation aggregation;
```
Comment (mariusoe) [inline inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManagerTest.java]: Not used

File: inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+    private final Supplier<Aggregation> aggregationSupplier = () -> aggregation;
+
+    private final TagKey[] commonTags = {TagKey.create("common-A"), TagKey.create("common-B")};
+
+
+    @InjectMocks
+    MeasuresAndViewsManager manager = new MeasuresAndViewsManager();
```
Comment (mariusoe) [inline inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManagerTest.java]: Remove one empty line above and below this attribute please

File: inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            verify(viewManager, times(3)).registerView(viewArg.capture());
+
+            List<String> names = viewArg.getAllValues().stream()
+                    .map(v -> v.getName().asString())
+                    .collect(Collectors.toList());
+            assertThat(names).containsExactlyInAnyOrder("viewA", "viewB", "viewC");
+
+        }
+    }
+
```
Comment (mariusoe) [inline inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/metrics/MeasuresAndViewsManagerTest.java]: Can you also some small test for `tryRecordingMeasurement` and verify that `updateMetricDefinitions` is not throwing any exception?

**Pull request merged**